src/dune/expander.ml

```diff
@@ -245,14 +262,17 @@ module Resolved_forms = struct
       mutable ddeps : Value.t list Build.t Pform.Expansion.Map.t
     }
 
-  let failures t = t.failures
-
   let lib_deps t = t.lib_deps
 
   let sdeps t = t.sdeps
 
   let ddeps t = t.ddeps
 
+  let prefix_failures t b =
+    match t.failures with
+    | [] -> b
+    | fail :: _ -> Build.O.( >>> ) (Build.fail fail) b
```

Comment:
This refactoring is useful to make sure that we use the same logic to prefix rules with the failure discovered in the expansion